File: module_me/src/main/java/com/ebook/me/mvvm/model/ModifyModel.java
package com.ebook.me.mvvm.model;

import android.app.Application;

import com.blankj.utilcode.util.SPUtils;
import com.ebook.api.service.UserService;
import com.ebook.api.RetrofitManager;
import com.ebook.api.dto.RespDTO;
import com.ebook.api.http.RxAdapter;
import com.ebook.common.event.KeyCode;
import com.ebook.common.mvvm.model.BaseModel;

import java.io.File;

import io.reactivex.Observable;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

public class ModifyModel extends BaseModel {
    private UserService userService;

    public ModifyModel(Application application) {
        super(application);
        userService = RetrofitManager.getInstance().getUserService();
    }

    /**
     * 修改昵称
     */
    @SuppressWarnings("unchecked")
    public Observable<RespDTO<Integer>> modifyNickname(String nickname) {
        String username = SPUtils.getInstance().getString(KeyCode.Login.SP_USERNAME);
        return userService.modifyNickname(RetrofitManager.getInstance().TOKEN, username, nickname)
                .compose(RxAdapter.schedulersTransformer())
                .compose(RxAdapter.exceptionTransformer());
    }

    /**
     * 修改头像
     *
     * @param path 头像路径
     * @return 返回服务器头像名称
     */
    @SuppressWarnings("unchecked")
    public Observable<RespDTO<String>> modifyProfiePhoto(String path) {
        String username = SPUtils.getInstance().getString(KeyCode.Login.SP_USERNAME);
        File file = new File(path);
        MultipartBody.Part body = MultipartBody.Part.createFormData("file", file.getName(), RequestBody.create(MediaType.parse("application/octet-stream"), file));
        return userService.modifyProfiePhoto(RetrofitManager.getInstance().TOKEN, username, body)
                .compose(RxAdapter.schedulersTransformer())
                .compose(RxAdapter.exceptionTransformer());
    }

}

File: module_me/src/main/java/com/ebook/me/mvvm/model/ModifyModel.kt
package com.ebook.me.mvvm.model

import android.app.Application
import android.net.Uri
import com.ebook.api.service.user.UserDataSource
import com.ebook.api.utils.CoroutineAdapter
import com.ebook.common.event.KeyCode
import com.ebook.common.util.SPUtil
import com.xrn1997.common.dto.RespDTO
import com.xrn1997.common.manager.RetrofitManager
import com.xrn1997.common.mvvm.model.BaseModel
import com.xrn1997.common.util.FileUtil
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ModifyModel @Inject constructor(
    application: Application,
    private val dataSource: UserDataSource
) : BaseModel(application) {

    /**
     * 修改昵称
     */
    suspend fun modifyNickname(nickname: String): Result<RespDTO<Int>> {
        val username = SPUtil.get(KeyCode.Login.SP_USERNAME, "")
        return CoroutineAdapter.safeApiCall { dataSource.modifyNickname(RetrofitManager.TOKEN, username, nickname)}
    }

    /**
     * 修改头像
     *
     * @param uri 头像路径
     * @return 返回服务器头像名称
     */
    @Throws(IOException::class)
    suspend fun modifyProfilePhoto(uri: Uri): Result<RespDTO<String>> {
        val username = SPUtil.get(KeyCode.Login.SP_USERNAME, "")

        val inputStream = mApplication.contentResolver.openInputStream(uri)
            ?: throw IOException("Unable to open InputStream for URI: $uri")

        inputStream.use { stream ->
            val requestBody = stream.readBytes()
                .toRequestBody("application/octet-stream".toMediaTypeOrNull())

            val body = MultipartBody.Part.createFormData(
                "file",
                FileUtil.getFileName(".jpg"),
                requestBody
            )
            return CoroutineAdapter.safeApiCall { dataSource.modifyProfilePhoto(RetrofitManager.TOKEN, username, body)}
        }
    }

}
